### PR TITLE
Generate missing syscalls headers programatically

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

**Comment by keszybz** [comment]
Tests:

- [x] amd64
- [x] i386
- [ ] ppc64le
- [ ] ppc64be
- [ ] s390
- [ ] s390x
- [ ] arm
- [ ] arm64
- [ ] various mips
- [ ] x32
- [ ] ???

**Review by bluca — APPROVED** [review]
I have not reviewed the generated files/headers, but the rest looks good and a solid improvement over status quo. LGTM

**Comment by keszybz** [comment]
I tested this on arches I have access to.

**Comment by mbiebl** [comment]
I like that, thanks @keszybz .

I can try to run that on the available Debian porter boxes. This can take a while though.
Let me know if you want to wait for that.

**Comment by keszybz** [comment]
I think the part that is most likely to be wrong is the template (in the .py file), incl. the mapping for file names to architectures. Other parts are mostly mechanic, so I don't expect trouble there.

>  Let me know if you want to wait for that.

I would prefer to merge without waiting.

**Comment by bluca** [comment]
test-hashmap fails - concidence?

```
378/544 test-hashmap                            TIMEOUT 90.09 s 

--- command ---
SYSTEMD_KBD_MODEL_MAP='/tmp/autopkgtest.uslCiL/build.0xs/systemd/src/locale/kbd-model-map' PATH='/tmp/autopkgtest.uslCiL/build.0xs/systemd/build-deb:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin' SYSTEMD_LANGUAGE_FALLBACK_MAP='/tmp/autopkgtest.uslCiL/build.0xs/systemd/src/locale/language-fallback-map' /tmp/autopkgtest.uslCiL/build.0xs/systemd/build-deb/test-hashmap
--- stderr ---
/************ test_hashmap_funcs ************/
/* test_hashmap_copy */
/* test_hashmap_get_strv */
/* test_hashmap_move_one */
/* test_hashmap_move */
/* test_hashmap_replace */
/* test_hashmap_update */
/* test_hashmap_put */
/* test_hashmap_remove */
/* test_hashmap_remove2 */
/* test_hashmap_remove_value */
/* test_hashmap_remove_and_put */
/* test_hashmap_remove_and_replace */
/* test_hashmap_ensure_allocated */
/* test_hashmap_foreach */
/* test_hashmap_foreach_key */
/* test_hashmap_contains */
/* test_hashmap_merge */
/* test_hashmap_isempty */
/* test_hashmap_get */
/* test_hashmap_get2 */
/* test_hashmap_size */
/* test_hashmap_many (slow) */
trivial_hashmap_ops 1048576 <= 1973790 * 0.8 = 1.57903e+06
test took 12.841241s
crippled_hashmap_ops 16384 <= 30840 * 0.8 = 24672
test took 1.253620s
/* test_hashmap_free (slow, 1048576 entries) */
string_hash_ops test took 12.454993s
custom_free_hash_ops test took 14.927069s
/* test_hashmap_free_with_destructor */
/* test_hashmap_first */
/* test_hashmap_first_key */
/* test_hashmap_steal_first_key */
/* test_hashmap_steal_first */
/* test_hashmap_clear_free_free */
/* test_hashmap_clear_free_with_destructor */
/* test_hashmap_reserve */
/* test_path_hashmap */
/* test_string_strv_hashmap */
/************ test_ordered_hashmap_funcs ************/
/* test_ordered_hashmap_copy */
/* test_ordered_hashmap_get_strv */
/* test_ordered_hashmap_move_one */
/* test_ordered_hashmap_move */
/* test_ordered_hashmap_replace */
/* test_ordered_hashmap_update */
/* test_ordered_hashmap_put */
/* test_ordered_hashmap_remove */
/* test_ordered_hashmap_remove2 */
/* test_ordered_hashmap_remove_value */
/* test_ordered_hashmap_remove_and_put */
/* test_ordered_hashmap_remove_and_replace */
/* test_ordered_hashmap_ensure_allocated */
/* test_ordered_hashmap_foreach */
/* test_ordered_hashmap_foreach_key */
/* test_ordered_hashmap_contains */
/* test_ordered_hashmap_merge */
/* test_ordered_hashmap_isempty */
/* test_ordered_hashmap_get */
/* test_ordered_hashmap_get2 */
/* test_ordered_hashmap_size */
/* test_ordered_hashmap_many (slow) */
trivial_ordered_hashmap_ops 1048576 <= 1342177 * 0.8 = 1.07374e+06
test took 20.292800s
crippled_ordered_hashmap_ops 16384 <= 20971 * 0.8 = 16776.8
test took 2.918083s
/* test_ordered_hashmap_free (slow, 1048576 entries) */
string_hash_ops test took 17.152628s
-------
```

**Comment by mbiebl** [comment]
mipsel and mips64el

```
(sid_mips64el-dchroot)biebl@eller:~/systemd$ ninja -C build
ninja: Entering directory `build'
[408/1577] Compiling C object src/shared/libsystemd-shared-247.a.p/seccomp-util.c.o
../src/shared/seccomp-util.c: In function 'seccomp_memory_deny_write_execute':
../src/shared/seccomp-util.c:1699:2: warning: #warning "Consider adding the right mmap() syscall definitions here!" [-Wcpp]
 1699 | #warning "Consider adding the right mmap() syscall definitions here!"
      |  ^~~~~~~
[1577/1577] Generating serial-getty@.service with a custom command (wrapped by meson to capture output)
ninja: Entering directory `build'
[1/2] Running all tests.


(sid_mipsel-dchroot)biebl@eller:~/systemd$ ninja -C build
ninja -C build test
ninja: Entering directory `build'
[404/1579] Compiling C object src/shared/libsystemd-shared-247.a.p/seccomp-util.c.o
../src/shared/seccomp-util.c: In function 'seccomp_memory_deny_write_execute':
../src/shared/seccomp-util.c:1699:2: warning: #warning "Consider adding the right mmap() syscall definitions here!" [-Wcpp]
 1699 | #warning "Consider adding the right mmap() syscall definitions here!"
      |  ^~~~~~~

```

**Comment by keszybz** [comment]
> [408/1577] Compiling C object src/shared/libsystemd-shared-247.a.p/seccomp-util.c.o
> ../src/shared/seccomp-util.c: In function 'seccomp_memory_deny_write_execute':
> ../src/shared/seccomp-util.c:1699:2: warning: #warning "Consider adding the right mmap() syscall definitions here!" [-Wcpp]
>  1699 | #warning "Consider adding the right mmap() syscall definitions here!"
>       |  ^~~~~~~

I don't think that's related. The code in seccomp-util.c has it's own set of conditionals, which my PR didn't touch at all.

**Comment by keszybz** [comment]
> test-hashmap fails - concidence?

Yeah, I think so. test-hashmap is known to fail occasionally.

Thank you all for reviews and testing. I'll merge this. Please remain vigilant.

**Comment by mbiebl** [comment]
@keszybz Fwiw, I think we should run `ninja -C build update-syscall-tables update-syscall-header` during build and diff that against a (non-)existing header. This way I can just upload the package to the Debian archive and will get the results from the build log without having to poke porters to run that command manually. WDYT?

**Comment by glaubitz** [comment]
@keszybz Verified on ia64 and ppc64, sparc64 so far. Works as expected. Thank you very much for fixing this!

**Comment by keszybz** [comment]
> @keszybz Fwiw, I think we should run `ninja -C build update-syscall-tables update-syscall-header` during build

The result of that command is architecture independent and should not change for a while (until the next kernel is released). So the version that is now stored in git master can be used.

> and diff that against a (non-)existing header. This way I can just upload the package to the Debian archive and will get the results from the build log for all architectures without having to poke individual porters to run that command manually. WDYT?

I don't understand. Just compiling on those architectures would be nice (since a major mismatch would most likely be exposed by the static asserts). Running the unittests would be even nicer.

**Comment by keszybz** [comment]
> @keszybz Verified on ia64 and ppc64, sparc64 so far. Works as expected. Thank you very much for fixing this!

What flavour of ppc64?

**Comment by glaubitz** [comment]
> > @keszybz Verified on ia64 and ppc64, sparc64 so far. Works as expected. Thank you very much for fixing this!
> 
> What flavour of ppc64?

64-bit big-endian PowerPC with a POWER5 baseline which is what we use for Debian's ```ppc64``` port.

The other ports we have are 32-bit big-endian PowerPC (no idea about the baseline right now) which is ```powerpc``` and 64-bit little-endian PowerPC which is ```ppc64el```.

FWIW, you can get access to various architectures in the GCC compile farm: https://gcc.gnu.org/wiki/CompileFarm

In fact, I tested ppc64 and sparc64 there.

Let me know if you want tests on any other architecture, I have access to machines of virtually every architecture supported in Debian.

File: src/systemctl/systemctl-list-units.c [hunk 1]
```diff
@@ -109,7 +109,7 @@ static int output_units_list(const UnitInfo *unit_infos, unsigned c) {
 
         (void) table_set_empty_string(table, "-");
 
-        for (const UnitInfo *u = unit_infos; unit_infos && u - unit_infos < c; u++) {
+        for (const UnitInfo *u = unit_infos; unit_infos && (unsigned)(u - unit_infos) < c; u++) {
```
Comment (poettering) [inline src/systemctl/systemctl-list-units.c:112]: should be size_t

**Comment by mbiebl** [comment]
Just wanted to shout out a big thanks here:
https://buildd.debian.org/status/package.php?p=systemd&suite=experimental

**Pull request merged**